src/journal/journal-send.c

```diff
@@ -73,27 +74,38 @@ _public_ int sd_journal_print(int priority, const char *format, ...) {
 }
 
 _public_ int sd_journal_printv(int priority, const char *format, va_list ap) {
-
-        /* FIXME: Instead of limiting things to LINE_MAX we could do a
-           C99 variable-length array on the stack here in a loop. */
-
-        char buffer[8 + LINE_MAX], p[STRLEN("PRIORITY=") + DECIMAL_STR_MAX(int) + 1];
+        char p[STRLEN("PRIORITY=") + DECIMAL_STR_MAX(int) + 1];
+        char sbuf[LINE_MAX + 8] = "MESSAGE=";
         struct iovec iov[2];
+        int len;
+        va_list aq;
+        char *buffer = sbuf;
 
         assert_return(priority >= 0, -EINVAL);
         assert_return(priority <= 7, -EINVAL);
         assert_return(format, -EINVAL);
 
         xsprintf(p, "PRIORITY=%i", priority & LOG_PRIMASK);
 
-        memcpy(buffer, "MESSAGE=", 8);
-        vsnprintf(buffer+8, sizeof(buffer) - 8, format, ap);
+        va_copy(aq, ap);
+        len = vsnprintf(buffer + 8, LINE_MAX, format, aq);
+        va_end(aq);
+
+        if (len > (int)LONG_LINE_MAX - 8)
+                return -ENOBUFS;
+
+        /* Allocate large buffer to accomodate big message */
+        if (len > LINE_MAX) {
+                buffer = alloca(len + 8);
+                memcpy(buffer, "MESSAGE=", 8);
+                vsnprintf(buffer + 8, len, format, ap);
```

Comment:
doesn't matter, but you could rewrite this as `vsnprintf(mempcpy(buffer, "MESSAGE=", 8), len, format, ap);`

Reply:
oh, and needs to be `len+1` here too, instead of `len`

Reply:
hmm, and I think it would be good to collect the return value of vsnprintf() here too, and then add an `assert()` that compares it with the original value, to ensure we are doing things right.

Reply:
vsnprintf(mempcpy(buffer, "MESSAGE=", 8), len, format, ap); doesn't look so clean so kept memcpy() in separate line

Reply:
> hmm, and I think it would be good to collect the return value of vsnprintf() here too, and then add an `assert()` that compares it with the original value, to ensure we are doing things right.

added assert_return()